docs/modules/ROOT/pages/adr/0040-adr-for-codey-git-ssh-access.adoc

```diff
@@ -0,0 +1,221 @@
+
+Findings show that cloud provider LoadBalancer products impose hard limits on the number of listeners per IP:
+
+* Cloudscale: up to 100 listeners per LoadBalancer
+* Exoscale: up to 10 listeners per LoadBalancer
+
+These limits make a single-IP for all tenants design infeasible at large scale.
+We must therefore introduce LoadBalancer sharding and evaluate which proxy technology provides the most reliable and maintainable control plane for that model.
+====
+
+== Context
+
+Codey users want a secure way to push/pull Git repositories using SSH keys.
+Because SSH is not SNI-aware, a single IP necessarily implies one unique TCP port per customer instance.
+The design must:
+
+* Work in Kubernetes/OpenShift environments,
+* Support dynamic onboarding/offboarding of tenants,
+* Avoid proxy restarts during updates,
+* Provide high availability,
+* Scale to hundreds of tenants within provider limits.
+
+== Considered Options
+
+=== Option 1: Per-tenant LoadBalancer (distinct IP per instance)
+
+Each Codey customer's Forgejo instance would expose its SSH Service directly as `type: LoadBalancer`, letting the cloud provider allocate a public IP and open TCP/22 for that instance.
+This is the most straightforward design and uses Kubernetes native Service abstraction without any shared proxying layer.
+Network traffic from the Internet terminates directly on each tenant's Forgejo pod, no intermediate routing or multiplexing is required.
+Operationally, this model scales linearly: every new customer gets one LoadBalancer and one public IP.
+
+.Pros:
+* Simple datapath
+* Simpler to operate
+* No lateral movement risk
+
+.Cons:
+* Potentially hundreds of IPs
+* Linear cost
+* Violates consolidation goal
+
+Status: Discarded
+
+
+=== Option 2: One LB + one IP + file-based TCP proxy (graceful reloads)
+
+A single shared LoadBalancer IP would forward all inbound SSH traffic to a TCP proxy Deployment (for example HAProxy).
+The proxy configuration file would define one `bind :PORT` and `server <tenant>` pair per customer instance.
+A custom controller would maintain that configuration file (for example a ConfigMap) and trigger a graceful reload whenever tenants are added or removed.
+During reloads, HAProxy can keep existing sessions alive but still incurs short pauses for new connections.
+
+.Pros: 
+* Simple architecture
+* Portable
+
+.Cons:
+* Reloads interrupt active sessions
+* Config management complexity
+
+Status: Discarded
+
+=== Option 3: Dynamic proxy with runtime API (HAProxy Data Plane API / Envoy xDS)
+
+Instead of editing files, the controller would communicate with the proxy at runtime:
+
+  - For **HAProxy**, through its REST-based Data Plane API to add or remove frontends/backends dynamically.
+  - For **Envoy**, through an xDS control-plane that updates listeners and clusters live.
```

Comment:
You could also check Traefik, which is quite capable in regard to dynamic configuration on Kubernetes.